ArjunaJTA/jta/classes/com/arjuna/ats/internal/jta/transaction/arjunacore/subordinate/SubordinateAtomicAction.java

```diff
@@ -242,6 +242,22 @@ public int doOnePhaseCommit ()
 	    if (status == ActionStatus.ABORT_ONLY || doBeforeCompletion())
 	    {
 	        status = super.End(true);
+
+	        switch (super.getHeuristicDecision())
+	        {
+	        // commit was processed correctly but some of the resources failed to finish
+	        // for 2PC this means no trouble as in case of a crash at this time
+	        // there is recovery which works with prepared transaction
+	        // for 1PC where there is nothing prepared and nothing stored at object store
+	        // we need to inform parent about error by setting up heuristic state
+	        case TwoPhaseOutcome.PREPARE_OK:
+	        case TwoPhaseOutcome.FINISH_OK:
+	            if (super.failedList != null && super.failedList.size() > 0) {
+	                return ActionStatus.H_COMMIT;
```

Comment:
Is it okay to skip the afterCompletion(H_COMMIT) call that was previously performed at line 252

Reply:
I took the same approach as it's in the `doCommit`: https://github.com/jbosstm/narayana/blob/5.9.5.Final/ArjunaJTA/jta/classes/com/arjuna/ats/internal/jta/transaction/arjunacore/subordinate/SubordinateAtomicAction.java#L170

Plus my understanding is - when the transaction ends with the heuristic outcome we can't say it was completed. Aka [the doc says](https://docs.oracle.com/javaee/7/api/javax/transaction/Synchronization.html#afterCompletion-int-) that the callback is called after transaction is rolled-back or committed. Here we are in *unspecified* state.

Reply:
The approach taken in `doCommit` for the heuristic outcome is:
https://github.com/jbosstm/narayana/blob/5.9.5.Final/ArjunaJTA/jta/classes/com/arjuna/ats/internal/jta/transaction/arjunacore/subordinate/SubordinateAtomicAction.java#L179
which does call afterCompletion.